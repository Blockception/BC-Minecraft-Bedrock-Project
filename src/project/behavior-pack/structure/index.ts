/*	Auto generated	*/

export * from "./process";
export * from "./structure";