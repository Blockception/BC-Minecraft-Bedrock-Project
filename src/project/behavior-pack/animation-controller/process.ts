import { Types } from "bc-minecraft-bedrock-types";
import * as internal from "../../../internal/behavior-pack/animation-controller";
import { Json } from "../../../internal/json";
import { Documentation, SMap, TextDocument } from "../../../types";
import { References } from "../../../types/references";
import { harvestMolang } from "../../molang/harvest";
import { AnimationController } from "./animation-controller";

/**
 *
 * @param doc
 * @returns
 */
export function Process(doc: TextDocument): AnimationController[] | undefined {
  const uri = doc.uri;
  const content = doc.getText();
  const imp = Json.To<internal.AnimationControllers>(doc);

  if (!internal.AnimationControllers.is(imp)) return undefined;

  const out: AnimationController[] = [];

  const container = imp.animation_controllers;
  const keys = Object.getOwnPropertyNames(container);

  for (let I = 0; I < keys.length; I++) {
    const id = keys[I];
    const controller = container[id];

    if (internal.AnimationController.is(controller)) {
      const item: AnimationController = {
        id: id,
        animations: References.empty(),
        documentation: Documentation.getDoc(doc, () => `BP Animation Controller: ${id}`),
        events: [],
        location: Types.Location.create(uri, content.indexOf(id)),
        molang: harvestMolang(content, controller),
      };

      SMap.forEach(controller.states, (state) => {
        getAnimations(state, item.animations);
        state.on_entry
          ?.filter((entry) => entry.startsWith("@s "))
          .map((entry) => entry.slice(3))
          .forEach((entry) => {
            if (!item.events.includes(entry)) item.events.push(entry);
          });
        state.on_exit
          ?.filter((entry) => entry.startsWith("@s "))
          .map((entry) => entry.slice(3))
          .forEach((entry) => {
            if (!item.events.includes(entry)) item.events.push(entry);
          });
      });

      out.push(item);
    }
  }

  return out;
}

/**
 *
 * @param state
 * @param receiver
 */
function getAnimations(state: internal.State, receiver: Pick<References, "using">): void {
  if (state.animations) Types.Conditional.forEach(state.animations, (reference) => receiver.using.push(reference));
}
