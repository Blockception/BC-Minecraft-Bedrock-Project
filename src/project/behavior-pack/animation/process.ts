import { Types } from "bc-minecraft-bedrock-types";
import { Json } from "../../../internal";
import * as Internal from "../../../internal/behavior-pack";
import { Documentation, SMap, TextDocument } from "../../../types";
import { harvestMolang } from "../../molang";
import { Animation } from "./animation";

/** */
export function Process(doc: TextDocument): Animation[] | undefined {
  const uri = doc.uri;
  const content = doc.getText();
  const imp = Json.To<Internal.Animations>(doc);

  if (!Internal.Animations.is(imp)) return undefined;

  const out: Animation[] = [];
  const container = imp.animations;
  const keys = Object.getOwnPropertyNames(container);

  for (let I = 0; I < keys.length; I++) {
    const id = keys[I];
    const anim = container[id];

    if (Internal.Animation.is(anim)) {
      const events: string[] = [];

      SMap.forEach(anim.timeline, (keyframe) => {
        if (typeof keyframe == "string") keyframe = [keyframe];
        keyframe
          .filter((entry) => entry.startsWith("@s "))
          .map((entry) => entry.slice(3))
          .forEach((entry) => {
            if (!events.includes(entry)) events.push(entry);
          });
      });

      out.push({
        events: events,
        id: id,
        location: Types.Location.create(uri, content.indexOf(id)),
        molang: harvestMolang(content, anim),
        documentation: Documentation.getDoc(
          doc,
          () => `BP Animation: \`${id}\`, loop: ${anim.loop ?? false}, length: ${anim.animation_length ?? "unknown"}`
        ),
      });
    }
  }

  return out;
}
