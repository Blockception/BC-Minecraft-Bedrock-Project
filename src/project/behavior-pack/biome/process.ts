import { Types } from "bc-minecraft-bedrock-types";
import * as Internal from "../../../internal/behavior-pack/biome";
import { Json } from "../../../internal/json";
import { Defined, Documentation, TextDocument } from "../../../types";
import { Biome } from "./biome";

/**
 *
 * @param doc
 * @returns
 */
export function Process(doc: TextDocument): Biome | undefined {
  const uri = doc.uri;
  const content = doc.getText();
  const biome = Json.To<Internal.Biome>(doc);

  if (!Internal.Biome.is(biome)) return undefined;

  const id = biome["minecraft:biome"].description.identifier;
  const tagComp = biome["minecraft:biome"]?.components["minecraft:tags"];

  return {
    id: id,
    documentation: Documentation.getDoc(doc, () => `Biome: ${id}`),
    location: Types.Location.create(uri, content.indexOf(id)),
    tags: Defined.wrap(Array.isArray(tagComp) ? tagComp : Array.isArray(tagComp.tags) ? tagComp.tags : []),
  };
}
