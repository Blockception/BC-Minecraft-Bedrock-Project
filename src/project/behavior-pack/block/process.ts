import { Types } from "bc-minecraft-bedrock-types";
import { Molang } from "bc-minecraft-molang";
import { Json } from "../../../internal";
import { Documentation, SMap, TextDocument } from "../../../types";
import { Block } from "./block";
import { BlockState } from "./block-state";

import * as Internal from "../../../internal/behavior-pack";

/**
 *
 * @param doc
 * @returns
 */
export function Process(doc: TextDocument): Block | undefined {
  const uri = doc.uri;
  const content = doc.getText();
  const imp = Json.To<Internal.Block>(doc);

  if (!Internal.Block.is(imp)) return undefined;

  const container = imp["minecraft:block"];
  const id = container.description.identifier;

  const out: Block = {
    id: id,
    location: Types.Location.create(uri, content.indexOf(id)),
    molang: Molang.MolangSet.harvest(container, content),
    states: [],
    documentation: Documentation.getDoc(doc, () => `BP Block: ${id}`),
  };

  const props = container.description.properties;
  if (props)
    SMap.forEach(props, (values, prop) => {
      const s = BlockState.create(prop, values);

      if (s) out.states.push(s);
    });

  return out;
}
