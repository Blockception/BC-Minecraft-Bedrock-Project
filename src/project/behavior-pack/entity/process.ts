import { Types } from "bc-minecraft-bedrock-types";
import { ComponentContainer } from "bc-minecraft-bedrock-types/lib/minecraft/components";
import { Molang } from "bc-minecraft-molang";
import * as internal from "../../../internal/behavior-pack/entity";
import { Json } from "../../../internal/json";
import { Documentation, SMap, TextDocument } from "../../../types";
import { Entity } from "./entity";
import { EntityProperty } from "./properties";
import { References } from "../../../types/references";

/**
 *
 * @param doc
 * @returns
 */
export function Process(doc: TextDocument): Entity | undefined {
  const uri = doc.uri;
  const content = doc.getText();
  const imp = Json.To<internal.Entity>(doc);

  if (!internal.Entity.is(imp)) return undefined;

  const container = imp["minecraft:entity"];
  const id = container.description.identifier;

  const out: Entity = {
    animations: References.empty(),
    documentation: Documentation.getDoc(doc, () => `BP Entity: ${id}`),
    events: [],
    families: [],
    groups: [],
    id: id,
    location: Types.Location.create(uri, content.indexOf(id)),
    molang: Molang.MolangSet.harvest(container, content),
    properties: [],
    runtime_identifier: "",
  };

  if (container.component_groups) {
    SMap.forEach(container.component_groups, (group, name) => {
      out.groups.push(name);
      getFamilies(group, out.families);
    });
  }
  if (container.events) SMap.forEach(container.events, (event, name) => out.events.push(name));

  //Animations
  if (container.description.animations) {
    SMap.forEach(container.description.animations, (anim, name) => {
      out.animations.defined.push(name);
      out.animations.using.push(anim);
    });
  }

  const runtime_identifier = container.description.runtime_identifier;
  if (runtime_identifier) out.runtime_identifier = runtime_identifier;

  if (container.description.properties) {
    for (const [name, property] of Object.entries(container.description.properties)) {
      out.properties.push({ name, ...property } as EntityProperty);
    }
  }

  getFamilies(container.components, out.families);

  return out;
}

function getFamilies(components: ComponentContainer, receiver: string[]) {
  const families = components["minecraft:type_family"];

  if (type_family.is(families)) {
    receiver.push(...families.family);
  }
}

interface type_family {
  family: string[];
}

namespace type_family {
  export function is(value: any): value is type_family {
    if (typeof value === "object" && Array.isArray(value.family)) {
      return true;
    }

    return false;
  }
}
