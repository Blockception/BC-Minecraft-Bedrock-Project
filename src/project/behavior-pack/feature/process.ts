import { Types } from "bc-minecraft-bedrock-types";
import * as Internal from "../../../internal/behavior-pack/feature";
import { Json } from "../../../internal/json";
import { Documentation, TextDocument } from "../../../types";
import { Feature } from "./feature";

/**
 *
 * @param doc
 * @returns
 */
export function process(doc: TextDocument): Feature | undefined {
  const uri = doc.uri;
  const content = doc.getText();
  const imp = Json.To<Internal.Feature>(doc);

  if (!Internal.Feature.is(imp)) return undefined;

  const key = Object.keys(imp).find((x) => !x.startsWith("format_version"));
  if (key === undefined) return;
  const container = imp[key as keyof Internal.Feature];
  if (container === undefined || typeof container === "string") return;
  const id = container.description.identifier;

  return {
    id: id,
    documentation: Documentation.getDoc(doc, () => `Feature: ${id}`),
    location: Types.Location.create(uri, content.indexOf(id)),
    type: key,
  };
}
