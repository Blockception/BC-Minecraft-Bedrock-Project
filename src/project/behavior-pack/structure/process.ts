import { Types } from "bc-minecraft-bedrock-types";
import { TextDocument } from "../../../types/TextDocument";
import { Structure } from "./structure";

/**
 *
 * @param doc
 * @returns
 */
export function Process(doc: TextDocument): Structure | undefined {
  const uri = doc.uri;
  let index = uri.indexOf("structures");

  if (index < 0) return undefined;
  index += 11;

  let id = uri.substring(index, uri.length).replace(/\\/g, "/");
  id = id.replace(".mcstructure", "");

  if (id.includes('/')) {
    id = '"' + id + '"';
  }

  const out: Structure = {
    id: id,
    location: Types.Location.create(uri, 0),
    documentation: `McStructure: ${id}`,
  };

  return out;
}
