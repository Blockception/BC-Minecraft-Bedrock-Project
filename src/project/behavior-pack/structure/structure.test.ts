import { Process } from "../../../../src/project/behavior-pack/structure/process";

describe("Structure", () => {
  const data: { uri: string; result: string }[] = [
    {
      uri: "F:\\Temp2\\world\\behavior_packs\\EW-BP\\structures\\empty\\air_1.mcstructure",
      result: '"empty/air_1"',
    },
    {
      uri: "F:/Temp2/world/behavior_packs/EW-BP/structures/empty/air_1.mcstructure",
      result: '"empty/air_1"',
    },
    {
      uri: "F:/Temp2/world/behavior_packs/EW-BP/structures/empty/temp/air_1.mcstructure",
      result: '"empty/temp/air_1"',
    },
  ];

  test.each(data)("$result from $uri", (item) => {
    const out = Process({ uri: item.uri, getText: () => "//example" });

    expect(out).toMatchSnapshot();
  });
});
