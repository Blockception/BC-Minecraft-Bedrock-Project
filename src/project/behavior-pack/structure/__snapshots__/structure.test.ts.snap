// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Structure "empty/air_1" from F:/Temp2/world/behavior_packs/EW-BP/structures/empty/air_1.mcstructure 1`] = `
{
  "documentation": "McStructure: "empty/air_1"",
  "id": ""empty/air_1"",
  "location": {
    "position": 0,
    "uri": "F:/Temp2/world/behavior_packs/EW-BP/structures/empty/air_1.mcstructure",
  },
}
`;

exports[`Structure "empty/air_1" from F:\\Temp2\\world\\behavior_packs\\EW-BP\\structures\\empty\\air_1.mcstructure 1`] = `
{
  "documentation": "McStructure: "empty/air_1"",
  "id": ""empty/air_1"",
  "location": {
    "position": 0,
    "uri": "F:\\Temp2\\world\\behavior_packs\\EW-BP\\structures\\empty\\air_1.mcstructure",
  },
}
`;

exports[`Structure "empty/temp/air_1" from F:/Temp2/world/behavior_packs/EW-BP/structures/empty/temp/air_1.mcstructure 1`] = `
{
  "documentation": "McStructure: "empty/temp/air_1"",
  "id": ""empty/temp/air_1"",
  "location": {
    "position": 0,
    "uri": "F:/Temp2/world/behavior_packs/EW-BP/structures/empty/temp/air_1.mcstructure",
  },
}
`;
